genie-common-internal/src/main/java/com/netflix/genie/common/internal/util/DynamicPropertiesMapCache.java

```diff
@@ -0,0 +1,142 @@
+/*
+ *
+ *  Copyright 2020 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.common.internal.util;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.EnumerablePropertySource;
+import org.springframework.core.env.Environment;
+import org.springframework.core.env.PropertySource;
+import org.springframework.core.env.PropertySources;
+import org.springframework.core.env.StandardEnvironment;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+/**
+ * Utility class that produces a map of properties and their values if they match a given pattern.
+ * The map is cached for a fixed amount of time to avoid re-computing for each call.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class DynamicPropertiesMapCache {
+
+    private static final String CACHE_KEY = "properties-map";
+    private final Environment environment;
+    private final Pattern propertiesNamesPattern;
+    private final LoadingCache<String, Map<String, String>> cache;
```

Comment:
Seems like the cache instead of a single key "properties-map" could instead of multiple keys which are the actual unique patterns?

Reply:
The downside of a single shared cache is that all users need to agree on a single eviction policy.
For example: all properties sets must be refreshed every 5 seconds.

Also requires more thinking in terms of concurrency if used and shared by multiple other components.